File: spoj/MNNITAR.cpp
#include <vector>
#include <iostream>
using namespace std;

typedef vector<vector<long long> > matrix;
#define MOD 1000000007

matrix zero(2,vector<long long>(2));
long long tc,n,f;

matrix pr(matrix a, matrix b){
	matrix ret = zero;
	for (int i=0; i<2; ++i)
		for (int j=0; j<2; ++j)
			for (int k=0; k<2; ++k)
				ret[i][j] = (ret[i][j] + a[i][k]*b[k][j])%(MOD-1);
	return ret;
}

matrix po(matrix a, long long k){
	matrix ret = zero;
	if (k==0){
		ret[0][0] = 1; ret[0][1] = 0;
		ret[1][0] = 0; ret[1][1] = 1;
		return ret;
	}
	ret = po(a,k/2);
	ret = pr(ret,ret);
	if (k%2) ret = pr(ret,a);
	return ret;
}

long long pm(long long a,long long b){
	if (b==0) return 1;
	long long ret = pm(a,b/2);
	ret = (ret*ret)%MOD;
	if (b%2) ret = (ret*a)%MOD;
	return ret;
}

int main(){
	matrix init(2,vector<long long>(2));
	init[0][0] = 0;
	init[0][1] = 1;
	init[1][0] = 1;
	init[1][1] = 1;
	cin >> tc;
	while (tc--){
		cin >> n;
		if (n==0){
			cout << "1\n";
			continue;
		}
		cout << pm(2LL,po(init,n)[1][0]) << endl;
	}
	return 0;
}

File: spoj/MNNITAR.pas
type matrix = array[1..2,1..2] of int64;
const init:matrix = ((0,1),(1,1));
const iden:matrix = ((1,0),(0,1));
const mmod:int64  = 1000000007;

function pr(a,b:matrix):matrix;
  var
    ret: matrix;
    i,j,k: longint;
  
  begin
    fillchar(ret,sizeof(ret),0);
    for i:=1 to 2 do
      for j:=1 to 2 do
        for k:=1 to 2 do ret[i][j] := (ret[i][j] + a[i][k]*b[k][j]) mod (mmod-1);
    exit(ret);
  end;

function po(a:matrix; b:int64):matrix;
  var
    ret:matrix;
  
  begin
    if (b=0) then exit(iden);
    ret := po(a,b div 2);
    ret := pr(ret,ret);
    if (b and 1>0) then ret := pr(ret,a);
    exit(ret);
  end;

function pm(a,b:int64):int64;
  var
    ret:int64;
  
  begin
    if (b=0) then exit(1);
    ret := pm(a,b div 2);
    ret := (ret*ret) mod mmod;
    if (b and 1>0) then ret := (ret*a) mod mmod;
    exit(ret);
  end;

var
  tc,kasus:longint;
  n:int64;
  now:matrix;

begin
  readln(tc);
  for kasus := 1 to tc do
    begin
      readln(n);
      if (n=0) then
        begin
          writeln(1);
          continue;
        end;
      now := po(init,n);
      writeln(pm(2,now[2][1]));
    end;
end.
